htmd/parameterization/dihedral.py

```diff
@@ -0,0 +1,548 @@
+        upper_bounds[nterms:2*nterms] = 360
+
+        # Set offset bounds
+        lower_bounds[-self.numDihedrals:] = -10
+        upper_bounds[-self.numDihedrals:] = 10
+
+        return lower_bounds, upper_bounds
+
+    def _objective(self, x, _):
+        """
+        Objective function for the parameter fitting.
+        """
+
+        k0, phi0 = np.reshape(x[:-self.numDihedrals], (2, -1, self.MAX_DIHEDRAL_MULTIPLICITY))
+        offset = x[-self.numDihedrals:]
+
+        n = np.arange(1, self.MAX_DIHEDRAL_MULTIPLICITY + 1)
+        phi0 = np.deg2rad(phi0)
+
+        all_energies = []
+        for i in range(self.numDihedrals):
+            phis = self._angle_values_rad[i]
+            energies = np.sum(k0[i]*(1 + np.cos(n*phis - phi0[i])), axis=(1, 2)) + offset[i]
+            all_energies.append(energies)
+
+        all_energies = np.concatenate(all_energies)
+        rmsd = np.sqrt(np.mean((all_energies - self._all_target_energies)**2))
+
+        return rmsd
+
+    def _paramsToVector(self, params):
+        """
+        Convert the parameter objects to a vector.
+        """
+
+        assert [param.n for param in params] == list(range(1, self.MAX_DIHEDRAL_MULTIPLICITY+1))*self.numDihedrals
+
+        vector = [param.k0 for param in params] + [param.phi0 for param in params] + [0]*self.numDihedrals
+        vector = np.array(vector)
+
+        return vector
+
+    def _vectorToParams(self, vector, params):
+        """
+        Copy results from the vector to the parameter objects.
+        """
+
+        nparams = len(params)
+        assert vector.size == 2*nparams + self.numDihedrals
+
+        for i, param in enumerate(params):
+            param.k0 = float(vector[i])
+            param.phi0 = float(vector[i + nparams])
+
+        return params
+
+    def _optimize_CRS2_LM(self, vector):
+        """
+        Controled random search with local mutations
+        """
+
+        # Create a global optimizer
+        opt = nlopt.opt(nlopt.GN_CRS2_LM, vector.size)
+        opt.set_min_objective(self._objective)
+        lower_bounds, upper_bounds = self._getBounds()
+        opt.set_lower_bounds(lower_bounds)
+        opt.set_upper_bounds(upper_bounds)
+        neval = 10000*opt.get_dimension()  # TODO allow to tune this parameter
+        opt.set_maxeval(neval)
+
+        # Optimize parameters
+        vector = opt.optimize(vector) # TODO check optimizer status
+        self.loss = opt.last_optimum_value()
+        assert self._objective(vector, None) == self.loss
+
+        # Create a lcoal optimizer
+        opt = nlopt.opt(nlopt.LN_BOBYQA, opt.get_dimension())
+        opt.set_min_objective(self._objective)
+        opt.set_lower_bounds(lower_bounds)
+        opt.set_upper_bounds(upper_bounds)
+        opt.set_xtol_rel(1e-3)
+        opt.set_maxeval(neval)
+        opt.set_initial_step(1e-3*(upper_bounds-lower_bounds))
+
+        # Optimize parameters
+        vector = opt.optimize(vector) # TODO check optimizer status
```

Comment:
Just some terms inconsistency between this optimization and the random_search one, perhaps on purpose, just checking: in one you use the vector, loss generic terms, while in the other you use best_vector, best_loss.